sdk/python/approzium/psycopg2.py

```diff
@@ -45,107 +85,128 @@ def intataddress(address):
+            logging.debug("ApproziumConn __init__")
+            kwargs.pop("async", None)
+            kwargs.pop("async_", None)
+            conn = super().__init__(*args, **kwargs, async_=1)
+            if self.dsn is None:
+                # connection is uninitalized due to an error
+                return
+            self._salt = None
+            self._hash_sent = False
+            if is_sync:
+                wait(self)
+                set_connection_sync(self)
+                self.autocommit = False
+
+        def poll(self):
+            status = libpq_PQstatus(self.pgconn_ptr)
+            if status == self.CONNECTION_AWAITING_RESPONSE and not self._salt:
+                logging.debug("reading salt")
+                self._salt = read_salt(self)
+                return psycopg2.extensions.POLL_WRITE
+            elif self._salt and not self._hash_sent:
+                logging.debug("sending hash")
+                dbhost = self.get_dsn_parameters()["host"]
+                dbuser = self.get_dsn_parameters()["user"]
+                hash = get_hash(
+                    dbhost, dbuser, self._salt, approzium.authenticator_addr
+                )
+                send_hash(self, hash)
+                self._hash_sent = True
+                return psycopg2.extensions.POLL_WRITE
+            else:
+                logging.debug("normal poll")
+                return super().poll()
 
+    return ApproziumConn
 
-def parse_dsn_args(dsn, appz_args):
-    psycopg_dsn = ""
-    for d in dsn.split():
-        param, value = d.split("=")
-        if param in appz_args:
-            if appz_args[param] is None:
-                appz_args[param] = value
-        else:
-            psycopg_dsn += " " + d
-
-    if appz_args["authenticator"] is None:
-        raise ValueError("Authenticator not specified")
-    return psycopg_dsn
-
-
-def connect(dsn="", authenticator=None, iam_arn=None, async=0, **psycopgkwargs):
-    appz_args = {"authenticator": authenticator, "iam_arn": iam_arn}
-    psycopg_dsn = parse_dsn_args(dsn, appz_args)
-    pgconn = psycopg2.connect(psycopg_dsn, **psycopgkwargs, async=1)
-    salt = advance_until_challenge(pgconn)
-    dbhost = pgconn.get_dsn_parameters()["host"]
-    dbuser = pgconn.get_dsn_parameters()["user"]
-    hash = get_hash(iam_arn, dbhost, dbuser, salt, appz_args["authenticator"])
-    logging.debug(f"salt: {salt}, hash: {hash}")
-    send_hash(pgconn, hash)
-    advance_until_end(pgconn)
-    if async == 0:
-        set_connection_sync(pgconn)
-    return pgconn
+
+def connect(dsn=None, connection_factory=None, cursor_factory=None, **kwargs):
+    is_sync = True
+    if kwargs.get("async", False):
+        is_sync = False
+    if kwargs.get("async_", False):
+        is_sync = False
+    # construct our approzium factory class on top of given connection factory class
+    factory = construct_approzium_conn(connection_factory, is_sync)
+    return pgconnect(dsn, factory, cursor_factory, **kwargs)
```

Comment:
We use Psycopg2 native functionality that allows us to introduce a custom class factory. This allows us to cleanly override the poll parameter. Granted, we still do some hacky stuff to be able to support creating sync connections.